Source/Core/DolphinQt2/CheatsManager.cpp

```diff
@@ -0,0 +1,340 @@
+    m_match_operation->addItem(option);
+
+  auto* group_box = new QGroupBox(tr("Type"));
+  auto* group_layout = new QHBoxLayout;
+  group_box->setLayout(group_layout);
+
+  m_match_decimal = new QRadioButton(tr("Decimal"));
+  m_match_hexadecimal = new QRadioButton(tr("Hexadecimal"));
+  m_match_octal = new QRadioButton(tr("Octal"));
+
+  group_layout->addWidget(m_match_decimal);
+  group_layout->addWidget(m_match_hexadecimal);
+  group_layout->addWidget(m_match_octal);
+
+  layout->addWidget(m_result_label);
+  layout->addWidget(m_match_length);
+  layout->addWidget(m_match_operation);
+  layout->addWidget(m_match_value);
+  layout->addWidget(group_box);
+  layout->addWidget(m_match_new);
+  layout->addWidget(m_match_next);
+  layout->addWidget(m_match_reset);
+
+  // Splitters
+  m_option_splitter = new QSplitter(Qt::Horizontal);
+  m_table_splitter = new QSplitter(Qt::Vertical);
+
+  m_table_splitter->addWidget(m_match_table);
+  m_table_splitter->addWidget(m_watch_table);
+
+  m_option_splitter->addWidget(m_table_splitter);
+  m_option_splitter->addWidget(options);
+
+  return m_option_splitter;
+}
+
+size_t CheatsManager::GetTypeSize() const
+{
+  switch (m_match_length->currentIndex())
+  {
+  case TYPE_BYTE_INDEX:
+    return sizeof(u8);
+  case TYPE_SHORT_INDEX:
+    return sizeof(u16);
+  case TYPE_INT_INDEX:
+    return sizeof(u32);
+  case TYPE_FLOAT_INDEX:
+    return sizeof(float);
+  case TYPE_DOUBLE_INDEX:
+    return sizeof(double);
+  default:
+    return m_match_value->text().toStdString().size();
+  }
+}
+
+template <typename T>
+bool Compare(u8* ptr, T value, int op)
+{
+  T mem_value = *reinterpret_cast<T*>(ptr);
+
+  switch (op)
+  {
+  case CMP_EQUAL_INDEX:
+    return mem_value == value;
+  case CMP_NOT_EQUAL_INDEX:
+    return mem_value != value;
+  case CMP_LESS_INDEX:
+    return mem_value < value;
+  case CMP_LESS_EQUAL_INDEX:
+    return mem_value <= mem_value;
+  case CMP_MORE_INDEX:
+    return value > mem_value;
+  case CMP_MORE_EQUAL_INDEX:
+    return value >= mem_value;
+  default:
+    return false;
+  }
+}
+
+bool CheatsManager::MatchesSearch(u8* ptr) const
+{
+  const auto text = m_match_value->text();
+  const int op = m_match_operation->currentIndex();
+
+  const int base =
+      (m_match_decimal->isChecked() ? 10 : (m_match_hexadecimal->isChecked() ? 16 : 8));
+
+  switch (m_match_length->currentIndex())
+  {
+  case TYPE_BYTE_INDEX:
+    return Compare<u8>(ptr, text.toUShort(nullptr, base) & 0xFF, op);
+  case TYPE_SHORT_INDEX:
+    return Compare<short>(ptr, text.toUShort(nullptr, base), op);
+  case TYPE_INT_INDEX:
+    return Compare<u32>(ptr, text.toUInt(nullptr, base), op);
+  case TYPE_FLOAT_INDEX:
+    return Compare<float>(ptr, text.toFloat(), op);
+  case TYPE_DOUBLE_INDEX:
+    return Compare<double>(ptr, text.toFloat(), op);
+
+  case TYPE_STRING_INDEX:
+  {
+    std::string mem_text(reinterpret_cast<char*>(ptr), text.toStdString().size());
```

Comment:
You can likely use `std::equal` on the chars instead of constructing a separate std::string to do the comparison

Reply:
How exactly would I go about doing that?

Reply:
If I'm understanding it right, it should be (for `CompareType::Equal`):

```
const std::string std_text = text.toStdString();
return std::equal(std_text.cbegin(), std_text.cend(),
                  reinterpret_cast<char*>(Memory::m_pRAM + addr - 0x80000000));
```

Reply:
Yup, that's what I had meant.